### PR TITLE
[doc] Add an OTBN introduction page.

### PR DESCRIPTION
**Pull request by jadephilipoom**
This page is intended to be a starting place for people with some level of technical background to understand what OTBN is and why it's designed the way it is. Most of the details are still on other pages, but I wanted something that will cover the main ideas and get people oriented and interested in learning more. I requested reviews from everyone I thought might have an opinion on documentation style and OTBN, but feel free to add anyone else who might have feedback if I left someone off!

Right now the benchmarks are missing RSA and X25519; I'll add those once https://github.com/lowRISC/opentitan/pull/20136 and https://github.com/lowRISC/opentitan/pull/20137 are merged, so they're properly reproducible. (edit: X25519 benchmark now added)

Because it's annoying to read docs pages as git diffs, here's a printout of the page as a PDF (the links don't work and the formatting is a bit weird, so if you want to see the non-text elements in all their glory you can always check out this branch and build the docs locally for the proper interactive version):
[otbn_intro.pdf](https://github.com/lowRISC/opentitan/files/13054061/otbn_intro.pdf)

### CODE REVIEW

**Comment by jwnrt** [comment]
> Because it's annoying to read docs pages as git diffs

FYI, GitHub's "rich diff" works quite good for the markdown here, not sure when they added that feature

https://github.com/lowRISC/opentitan/pull/20140/files?short_path=5538186#diff-5538186acef0e9ec434b91d4882d55777e253d7f6106e4963bd5277317cd1d6e

**Comment by pamaury** [comment]
I know very little about the OTBN but I just want to say this is really great that you took the time to write this introduction, it's definitely very useful!

**Comment by jadephilipoom** [comment]
> FYI, GitHub's "rich diff" works quite good for the markdown here, not sure when they added that feature

Oh, this is cool, didn't know about that feature!

**Review by jwnrt — APPROVED** [review]
I'm also not an OTBN expert, but this reads very clearly and makes for great documentation. Thanks a lot @jadephilipoom.

**Review by hcallahan-lowrisc — APPROVED** [review]
This is great! Thanks @jadephilipoom for the high quality docs. This really tells a compelling story about what OTBN is and how to use it. The pictures are ace. I've left a bunch of minor suggestions, mostly grammar just to make the prose a little less informal in places.

One more overarching comment I have is that this doc describes OTBN in the context of the Ibex processor only. While this is the only meaningful context in OT, it certainly could be used differently elsewhere. I don't think it should be addressed in this PR, but longer-term it might make more sense to write this sort of content with language geared towards a more-general application (Ibex -> System Processor) or something... . Although I think this doc is great for OpenTitan users specifically. Maybe there should be two docs, an OT-specific one (e.g. integration guide) and a more-general one? I'm not sure what it would look like right now. Just some vague ideas :)

**Review by moidx — APPROVED** [review]
Awesome job @jadephilipoom! This document will be extremely helpful in showcasing OTBN to the extended community.

**Comment by jadephilipoom** [comment]
Thanks for all the detailed feedback, @hcallahan-lowrisc ! The latest push should have addressed all of it :slightly_smiling_face: In some places I rephrased the sentences a little differently than the suggestion just to match my own writing style -- hopefully that's OK.

> One more overarching comment I have is that this doc describes OTBN in the context of the Ibex processor only. While this is the only meaningful context in OT, it certainly could be used differently elsewhere.

Good point -- I added a sentence to the beginning that says: "The OTBN hardware block could also run as part of a different system and interact with a different main processor, but this page will focus on the OpenTitan context."

File: hw/ip/otbn/doc/otbn_intro.md [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+### Multiplying big numbers
+
+OTBN's multiplication instructions are `bn.mulqacc`, `bn.mulqacc.so`, and `bn.mulqacc.wo`.
+All of them do roughly the same thing:
+- they perform a 64x64-bit multiplication (the `q` in `mulqacc` is for "quarter-word")
+- they accumulate the 128-bit product into a special 256-bit special register called `ACC`
+
+The `.wo` variant then copies the entire accumulator value to another wide register.
+The `.so` variant writes the low 128 bits of the accumulator into a register and then shifts the accumulator 128 bits.
+All variants accept an offset argument, so the product can be added to the accumulator with a shift of 0, 64, or 192 bits.
```
Comment (sarranz) [inline hw/ip/otbn/doc/otbn_intro.md:58]: Isn't 128 also a valid shift?

**Review by sarranz — APPROVED** [review]

**Review by GregAC — APPROVED** [review]
Great work @jadephilipoom! This is fantastic addition to the documentation.

File: hw/ip/otbn/doc/otbn_intro.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+The following diagram illustrates how OTBN and the main processor interact:
+
+![Diagram showing OTBN and Ibex interaction](otbn_operation.svg)
+
+OTBN has its own data and instruction memories, distinct from the memories of the Ibex main processor.
+To run an OTBN program, Ibex loads the program into OTBN's instruction memory (and data memory if the program contains stored constants), then writes the input data into the data memory and flips the "execute" bit.
+When OTBN is done, it sends an interrupt to Ibex, which reads back the results from data memory.
+See [below](#an-example-program) for an example of a standalone OTBN program.
+
+OTBN protects its data from Ibex in various ways:
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/doc/otbn_intro.md:21]: I think an extra bullet here mentioning Key Sideloading from the isolated keymgr connection as part of the security model would be helpful.

File: hw/ip/otbn/doc/otbn_intro.md [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+For example, ECDSA-P256 has the entrypoint [p256_ecdsa](https://github.com/lowRISC/opentitan/blob/7528f848214589e837ce3b0dacac8385c458b772/sw/otbn/crypto/p256_ecdsa.s) for production code, and [p256_ecdsa_sca](https://github.com/lowRISC/opentitan/blob/7528f848214589e837ce3b0dacac8385c458b772/sw/otbn/crypto/p256_ecdsa_sca.s) for side channel analysis.
+They call the same underlying library, but the SCA-specific entrypoint allows the caller to select the per-signature secret value `k`, which is always randomly generated in the production-code version.
+We can then determine if information about `k` is leaking by trying different known values and seeing if values of `k` that are similar in a certain way have similarities in their traces.
+For example, during development we were able to fix a bug in our original implementation that leaked information about the number of leading zeroes in `k`.
+
+## Modeling and formal methods
+
+OTBN is well-suited to modeling because of its relatively simple ISA (52 instructions) and predictable timing behavior.
+This means we can relatively easily simulate OTBN's behavior in software and in formal methods tools.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/doc/otbn_intro.md:330]:
```suggestion
This means we can easily simulate OTBN's behavior in software and in formal methods tools.
```
relatively on the line before :)

File: hw/ip/otbn/doc/otbn_intro.md [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+| SHA-512 (1 block) | 3971 | 875b3a | sha512_test | yes |
+
+A few notes:
+- Because some OTBN code is still under development, these cycle counts are expected to change a bit as we optimize the code and add hardening countermeasures against fault injection and power/EM side-channel attacks.
+- Some of these benchmarks include significant overhead from these countermeasures (for example, we run the inner loop of P-256 scalar multiplication 320 times instead of 256), but in OpenTitan's threat model the price is worthwhile.
+- For non-constant-time code, due to the nature of the OTBN benchmarks, it is currently difficult to run multiple tests, so the numbers above reflect only one test each and should be treated as a rough estimate.
+
+### Benchmark reproduction
+
+To reproduce these benchmarks yourself, checkout out the specified commit from OpenTitan, then run the OTBN simulator directly on the specified programs.
```
Comment (hcallahan-lowrisc) [inline hw/ip/otbn/doc/otbn_intro.md:295]:
```suggestion
To reproduce these benchmarks yourself, checkout the specified commit from OpenTitan, then run the OTBN simulator directly on the specified programs.
```

**Review by hcallahan-lowrisc — COMMENTED** [review]
> The latest push should have addressed all of it 🙂 In some places I rephrased the sentences a little differently than the suggestion just to match my own writing style -- hopefully that's OK.

Great, thanks for addressing my suggestions. Feel free to use/tweak/ignore as you see fit!

File: hw/ip/otbn/doc/otbn_intro.md (hunk 1 above)
Comment (jadephilipoom) [reply hw/ip/otbn/doc/otbn_intro.md:58]: Yes, thanks for catching that!

**Comment by jadephilipoom** [comment]
Latest commit adds the RSA benchmarks now that https://github.com/lowRISC/opentitan/pull/20136 is merged. These numbers should improve significantly once https://github.com/lowRISC/opentitan/pull/19755 and a couple planned follow-ups are merged; the RSA code is functional but slow at the moment.

**Comment by jadephilipoom** [comment]
I'll merge now, but if anyone has post-merge comments feel free to post them here and I'll make a follow-up!

**Pull request merged**